Collect previous LDM work and document step-by-step usage instructions

frankhereford opened the pull request:
# Reviewing

This PR contains materials and instructions which are intended to take the VZ DB out of production and entirely transform it into a LDM enabled database, including both fact and dimension/lookup tables. It's based off the previous LDM PR, which provides the framework for the lookup table system.

Running these instructions end-to-end will take 5+ hours. I've attempted to make each step idempotent, but sometimes you need to pair up operations to be able to do certain steps over and over. 

I have attempted to make the testing instructions as clear as possible and make no assumptions nor requirements on the reviewer to put together environments or commands. This opens up review to be conducted at whatever level of detail the reviewer wants to give it. 

There are multiple scripts that will be used throughout the process. It's a lot to review, and I am incredibly happy to provide more information about any part of this. Please don't hesitate to ask. I'll respond promptly and in detail.

# Associated issues

https://github.com/cityofaustin/atd-data-tech/issues/13715

# LDM Schema naming

  - I'm proposing we use these:
    - `cris_facts`
    - `vz_facts`
    - `cris_lookup`
    - `vz_lookup`
    - `public` Both of the above pairs have their `union`ed tables appear in public, to emulate the tables we intend for reading data.
    - `import_...` Ephemeral schema that only exist while the ETL is running
    - `production_fact_tables` Ephemeral schema that exists during the launch of the LDM

# Testing

## Get production VZ data

- Replicate production into `atd_vz_data`, using the stack tool

```bash
./vision-zero replicate-db
```

## Add new/missing columns

- Let's now add the columns which were added in the 2023 CRIS schema or were just missing from the original VZDB.

```
docker exec -i \
-e PGHOST=localhost \
-e PGDATABASE=atd_vz_data \
-e PGPASSWORD=visionzero \
-e PGUSER=visionzero \
visionzero-postgis \
psql --set ON_ERROR_STOP=on \
-c "alter table public.atd_txdot_units add column rpt_autonomous_unit_id integer default null;"

docker exec -i \
-e PGHOST=localhost \
-e PGDATABASE=atd_vz_data \
-e PGPASSWORD=visionzero \
-e PGUSER=visionzero \
visionzero-postgis \
psql --set ON_ERROR_STOP=on \
-c "alter table public.atd_txdot_units add column rpt_autonomous_level_engaged_id integer default null;"

docker exec -i \
-e PGHOST=localhost \
-e PGDATABASE=atd_vz_data \
-e PGPASSWORD=visionzero \
-e PGUSER=visionzero \
visionzero-postgis \
psql --set ON_ERROR_STOP=on \
-c "alter table public.atd_txdot_primaryperson add column drvr_lic_state_id integer default null;"
```

## Move VZ data into a holding DB

- Create temp holding DB for production data, `copy_of_production`

```bash
docker exec -i \
-e PGHOST=localhost \
-e PGDATABASE=atd_vz_data \
-e PGPASSWORD=visionzero \
-e PGUSER=visionzero \
visionzero-postgis \
psql --set ON_ERROR_STOP=on \
-c "drop database if exists copy_of_production with (force);"

docker exec -i \
-e PGHOST=localhost \
-e PGDATABASE=atd_vz_data \
-e PGPASSWORD=visionzero \
-e PGUSER=visionzero \
visionzero-postgis \
psql --set ON_ERROR_STOP=on \
-c "create database copy_of_production;"
```

- Copy `atd_vz_data` into your `copy_of_production` DB

```bash
docker exec -i \
-e PGHOST=localhost \
-e PGDATABASE=atd_vz_data \
-e PGPASSWORD=visionzero \
-e PGUSER=visionzero \
visionzero-postgis \
pg_dump | \
perl -pe 's/^(SELECT pg_catalog\.set_config)/-- $1/g' | \
docker exec -i \
-e PGHOST=localhost \
-e PGDATABASE=copy_of_production \
-e PGPASSWORD=visionzero \
-e PGUSER=visionzero \
visionzero-postgis \
psql --set ON_ERROR_STOP=on 
```
## Create output DB

- Drop and create the database which is going to house the finished product

```bash
docker exec -i \
-e PGHOST=localhost \
-e PGDATABASE=postgres \
-e PGPASSWORD=visionzero \
-e PGUSER=visionzero \
visionzero-postgis \
psql --set ON_ERROR_STOP=on \
-c "drop database if exists atd_vz_data with (force);"

docker exec -i \
-e PGHOST=localhost \
-e PGDATABASE=postgres \
-e PGPASSWORD=visionzero \
-e PGUSER=visionzero \
visionzero-postgis \
psql --set ON_ERROR_STOP=on \
-c "create database atd_vz_data;"

docker exec -i \
-e PGHOST=localhost \
-e PGDATABASE=atd_vz_data \
-e PGPASSWORD=visionzero \
-e PGUSER=visionzero \
visionzero-postgis \
psql --set ON_ERROR_STOP=on \
-c "create extension postgis;"
```

## Build up LDM enabled DB structure

- Create the CRIS schema for fact tables

```bash
docker exec -i \
-e PGHOST=localhost \
-e PGDATABASE=atd_vz_data \
-e PGPASSWORD=visionzero \
-e PGUSER=visionzero \
visionzero-postgis \
psql --set ON_ERROR_STOP=on \
-c "drop schema if exists cris_facts cascade;"

docker exec -i \
-e PGHOST=localhost \
-e PGDATABASE=atd_vz_data \
-e PGPASSWORD=visionzero \
-e PGUSER=visionzero \
visionzero-postgis \
psql --set ON_ERROR_STOP=on \
-c "create schema cris_facts;"
```

## Prepare to pull over fact tables

- We’re going to need some functions to come over, so let’s do that for the following
    - `public.is_crash_in_austin_full_purpose()`
    - `public.find_council_district()`
- ⚠️ The PRs related to this from Rose & Mike may change how this works, and it’ll be during the merge process that this gets reconciled. This is a thing to consider and review and possibly discuss in the standup meeting.
    - These have to come over now because they are used in the computed fields, and therefore are part of the table definition itself.

```bash
# pull over public.is_crash_in_austin_full_purpose()
docker exec -i \
-e PGHOST=localhost \
-e PGDATABASE=copy_of_production \
-e PGPASSWORD=visionzero \
-e PGUSER=visionzero \
visionzero-postgis \
psql --tuples-only --no-align --set ON_ERROR_STOP=on  \
-c "SELECT pg_catalog.pg_get_functiondef(pg_proc.oid) 
FROM pg_catalog.pg_proc
JOIN pg_catalog.pg_namespace ON (pg_proc.pronamespace = pg_namespace.oid)
WHERE pg_namespace.nspname = 'public' AND pg_proc.proname = 'is_crash_in_austin_full_purpose'" | \
docker exec -i \
-e PGHOST=localhost \
-e PGDATABASE=atd_vz_data \
-e PGPASSWORD=visionzero \
-e PGUSER=visionzero \
visionzero-postgis \
psql --set ON_ERROR_STOP=on 

docker exec -i \
-e PGHOST=localhost \
-e PGDATABASE=copy_of_production \
-e PGPASSWORD=visionzero \
-e PGUSER=visionzero \
visionzero-postgis \
psql --tuples-only --no-align --set ON_ERROR_STOP=on  \
-c "SELECT pg_catalog.pg_get_functiondef(pg_proc.oid) 
FROM pg_catalog.pg_proc
JOIN pg_catalog.pg_namespace ON (pg_proc.pronamespace = pg_namespace.oid)
WHERE pg_namespace.nspname = 'public' AND pg_proc.proname = 'find_council_district'" | \
docker exec -i \
-e PGHOST=localhost \
-e PGDATABASE=atd_vz_data \
-e PGPASSWORD=visionzero \
-e PGUSER=visionzero \
visionzero-postgis \
psql --set ON_ERROR_STOP=on 
```

## Create empty fact tables for CRIS data

- Pull over each fact table. Each is a little different, so this is not done in a loop. You could do it in a loop, but it’d get real tough trying to keep track of what is what `pg_dump` output. This way, you can inspect the particular `pg_dump` output easily but piping into a text buffer, `vim -`, `more` or `less`, instead of the `psql` client. 🥁 😜

```bash
## Crashes table

docker exec -i \
-e PGHOST=localhost \
-e PGDATABASE=atd_vz_data \
-e PGPASSWORD=visionzero \
-e PGUSER=visionzero \
visionzero-postgis \
psql --set ON_ERROR_STOP=on \
-c "drop table if exists cris_facts.atd_txdot_crashes cascade;"

docker exec -i \
-e PGHOST=localhost \
-e PGDATABASE=copy_of_production \
-e PGPASSWORD=visionzero \
-e PGUSER=visionzero \
visionzero-postgis \
pg_dump -t public.atd_txdot_crashes \
--no-owner --no-comments --no-acl --schema-only --disable-triggers | \
perl -pe 's/public.atd_txdot_crashes/cris_facts.atd_txdot_crashes/g' | \
perl -pe 's/^(CREATE TRIGGER.*)/-- $1/g' | \
perl -pe 's/^(SELECT pg_catalog\.set_config)/-- $1/g' | \
docker exec -i \
-e PGHOST=localhost \
-e PGDATABASE=atd_vz_data \
-e PGPASSWORD=visionzero \
-e PGUSER=visionzero \
visionzero-postgis \
psql --set ON_ERROR_STOP=on

## Units table

docker exec -i \
-e PGHOST=localhost \
-e PGDATABASE=atd_vz_data \
-e PGPASSWORD=visionzero \
-e PGUSER=visionzero \
visionzero-postgis \
psql --set ON_ERROR_STOP=on \
-c "drop table if exists cris_facts.atd_txdot_units cascade;"

docker exec -i \
-e PGHOST=localhost \
-e PGDATABASE=copy_of_production \
-e PGPASSWORD=visionzero \
-e PGUSER=visionzero \
visionzero-postgis \
pg_dump -t public.atd_txdot_units \
--no-owner --no-comments --no-acl --schema-only --disable-triggers | \
perl -pe 's/public.atd_txdot_crashes/cris_facts.atd_txdot_crashes/g' | \
perl -pe 's/public.atd_txdot_units/cris_facts.atd_txdot_units/g' | \
perl -pe 's/^(CREATE TRIGGER.*)/-- $1/g' | \
perl -pe 's/^(SELECT pg_catalog\.set_config)/-- $1/g' | \
docker exec -i \
-e PGHOST=localhost \
-e PGDATABASE=atd_vz_data \
-e PGPASSWORD=visionzero \
-e PGUSER=visionzero \
visionzero-postgis \
psql --set ON_ERROR_STOP=on

## Primary Persons

docker exec -i \
-e PGHOST=localhost \
-e PGDATABASE=atd_vz_data \
-e PGPASSWORD=visionzero \
-e PGUSER=visionzero \
visionzero-postgis \
psql --set ON_ERROR_STOP=on \
-c "drop table if exists cris_facts.atd_txdot_primaryperson cascade;"

docker exec -i \
-e PGHOST=localhost \
-e PGDATABASE=copy_of_production \
-e PGPASSWORD=visionzero \
-e PGUSER=visionzero \
visionzero-postgis \
pg_dump -t public.atd_txdot_primaryperson \
--no-owner --no-comments --no-acl --schema-only --disable-triggers | \
perl -pe 's/public.atd_txdot_crashes/cris_facts.atd_txdot_crashes/g' | \
perl -pe 's/public.atd_txdot_units/cris_facts.atd_txdot_units/g' | \
perl -pe 's/public.atd_txdot_primaryperson/cris_facts.atd_txdot_primaryperson/g' | \
perl -pe 's/^(CREATE TRIGGER.*)/-- $1/g' | \
perl -pe 's/^(SELECT pg_catalog\.set_config)/-- $1/g' | \
docker exec -i \
-e PGHOST=localhost \
-e PGDATABASE=atd_vz_data \
-e PGPASSWORD=visionzero \
-e PGUSER=visionzero \
visionzero-postgis \
psql --set ON_ERROR_STOP=on

# and Persons

docker exec -i \
-e PGHOST=localhost \
-e PGDATABASE=copy_of_production \
-e PGPASSWORD=visionzero \
-e PGUSER=visionzero \
visionzero-postgis \
psql --set ON_ERROR_STOP=on \
-c "alter table atd_txdot_person drop constraint if exists atd_txdot_person_prsn_injry_sev_id_fkey;"

docker exec -i \
-e PGHOST=localhost \
-e PGDATABASE=atd_vz_data \
-e PGPASSWORD=visionzero \
-e PGUSER=visionzero \
visionzero-postgis \
psql --set ON_ERROR_STOP=on \
-c "drop table if exists cris_facts.atd_txdot_person cascade;"

docker exec -i \
-e PGHOST=localhost \
-e PGDATABASE=copy_of_production \
-e PGPASSWORD=visionzero \
-e PGUSER=visionzero \
visionzero-postgis \
pg_dump -t public.atd_txdot_person \
--no-owner --no-comments --no-acl --schema-only --disable-triggers | \
perl -pe 's/^(SELECT pg_catalog\.set_config)/-- $1/g' | \
perl -pe 's/public.atd_txdot_crashes/cris_facts.atd_txdot_crashes/g' | \
perl -pe 's/public.atd_txdot_units/cris_facts.atd_txdot_units/g' | \
perl -pe 's/public.atd_txdot_person/cris_facts.atd_txdot_person/g' | \
perl -pe 's/^(CREATE TRIGGER.*)/-- $1/g' | \
perl -pe 's/(.*DISABLE TRIGGER.*)/-- $1/g' | \
docker exec -i \
-e PGHOST=localhost \
-e PGDATABASE=atd_vz_data \
-e PGPASSWORD=visionzero \
-e PGUSER=visionzero \
visionzero-postgis \
psql --set ON_ERROR_STOP=on

```

## Create empty VZ fact tables

- Now that we have the CRIS tables, albeit empty, we need the VZ tables, also empty. Let’s create a schema for them and then create the empty tables.

```bash
docker exec -i \
-e PGHOST=localhost \
-e PGDATABASE=atd_vz_data \
-e PGPASSWORD=visionzero \
-e PGUSER=visionzero \
visionzero-postgis \
psql --set ON_ERROR_STOP=on \
-c "drop schema if exists vz_facts cascade;"

docker exec -i \
-e PGHOST=localhost \
-e PGDATABASE=atd_vz_data \
-e PGPASSWORD=visionzero \
-e PGUSER=visionzero \
visionzero-postgis \
psql --set ON_ERROR_STOP=on \
-c "create schema vz_facts;"
```

```bash
## Crashes table

docker exec -i \
-e PGHOST=localhost \
-e PGDATABASE=atd_vz_data \
-e PGPASSWORD=visionzero \
-e PGUSER=visionzero \
visionzero-postgis \
psql --set ON_ERROR_STOP=on \
-c "drop table if exists vz_facts.atd_txdot_crashes cascade;"

docker exec -i \
-e PGHOST=localhost \
-e PGDATABASE=copy_of_production \
-e PGPASSWORD=visionzero \
-e PGUSER=visionzero \
visionzero-postgis \
pg_dump -t public.atd_txdot_crashes \
--no-owner --no-comments --no-acl --schema-only --disable-triggers | \
perl -pe 's/public.atd_txdot_crashes/vz_facts.atd_txdot_crashes/g' | \
perl -pe 's/^(CREATE TRIGGER.*)/-- $1/g' | \
perl -pe 's/^(SELECT pg_catalog\.set_config)/-- $1/g' | \
docker exec -i \
-e PGHOST=localhost \
-e PGDATABASE=atd_vz_data \
-e PGPASSWORD=visionzero \
-e PGUSER=visionzero \
visionzero-postgis \
psql --set ON_ERROR_STOP=on

## Units table

docker exec -i \
-e PGHOST=localhost \
-e PGDATABASE=atd_vz_data \
-e PGPASSWORD=visionzero \
-e PGUSER=visionzero \
visionzero-postgis \
psql --set ON_ERROR_STOP=on \
-c "drop table if exists vz_facts.atd_txdot_units cascade;"

docker exec -i \
-e PGHOST=localhost \
-e PGDATABASE=copy_of_production \
-e PGPASSWORD=visionzero \
-e PGUSER=visionzero \
visionzero-postgis \
pg_dump -t public.atd_txdot_units \
--no-owner --no-comments --no-acl --schema-only --disable-triggers | \
perl -pe 's/public.atd_txdot_crashes/vz_facts.atd_txdot_crashes/g' | \
perl -pe 's/public.atd_txdot_units/vz_facts.atd_txdot_units/g' | \
perl -pe 's/^(CREATE TRIGGER.*)/-- $1/g' | \
perl -pe 's/^(SELECT pg_catalog\.set_config)/-- $1/g' | \
docker exec -i \
-e PGHOST=localhost \
-e PGDATABASE=atd_vz_data \
-e PGPASSWORD=visionzero \
-e PGUSER=visionzero \
visionzero-postgis \
psql --set ON_ERROR_STOP=on

## Primary Persons

docker exec -i \
-e PGHOST=localhost \
-e PGDATABASE=atd_vz_data \
-e PGPASSWORD=visionzero \
-e PGUSER=visionzero \
visionzero-postgis \
psql --set ON_ERROR_STOP=on \
-c "drop table if exists vz_facts.atd_txdot_primaryperson cascade;"

docker exec -i \
-e PGHOST=localhost \
-e PGDATABASE=copy_of_production \
-e PGPASSWORD=visionzero \
-e PGUSER=visionzero \
visionzero-postgis \
pg_dump -t public.atd_txdot_primaryperson \
--no-owner --no-comments --no-acl --schema-only --disable-triggers | \
perl -pe 's/public.atd_txdot_crashes/vz_facts.atd_txdot_crashes/g' | \
perl -pe 's/public.atd_txdot_units/vz_facts.atd_txdot_units/g' | \
perl -pe 's/public.atd_txdot_primaryperson/vz_facts.atd_txdot_primaryperson/g' | \
perl -pe 's/^(CREATE TRIGGER.*)/-- $1/g' | \
perl -pe 's/^(SELECT pg_catalog\.set_config)/-- $1/g' | \
docker exec -i \
-e PGHOST=localhost \
-e PGDATABASE=atd_vz_data \
-e PGPASSWORD=visionzero \
-e PGUSER=visionzero \
visionzero-postgis \
psql --set ON_ERROR_STOP=on

# and Persons

docker exec -i \
-e PGHOST=localhost \
-e PGDATABASE=atd_vz_data \
-e PGPASSWORD=visionzero \
-e PGUSER=visionzero \
visionzero-postgis \
psql --set ON_ERROR_STOP=on \
-c "drop table if exists vz_facts.atd_txdot_person cascade;"

docker exec -i \
-e PGHOST=localhost \
-e PGDATABASE=copy_of_production \
-e PGPASSWORD=visionzero \
-e PGUSER=visionzero \
visionzero-postgis \
pg_dump -t public.atd_txdot_person \
--no-owner --no-comments --no-acl --schema-only --disable-triggers | \
perl -pe 's/^(SELECT pg_catalog\.set_config)/-- $1/g' | \
perl -pe 's/public.atd_txdot_crashes/vz_facts.atd_txdot_crashes/g' | \
perl -pe 's/public.atd_txdot_units/vz_facts.atd_txdot_units/g' | \
perl -pe 's/public.atd_txdot_person/vz_facts.atd_txdot_person/g' | \
perl -pe 's/^(CREATE TRIGGER.*)/-- $1/g' | \
perl -pe 's/(.*DISABLE TRIGGER.*)/-- $1/g' | \
docker exec -i \
-e PGHOST=localhost \
-e PGDATABASE=atd_vz_data \
-e PGPASSWORD=visionzero \
-e PGUSER=visionzero \
visionzero-postgis \
psql --set ON_ERROR_STOP=on
```

## Populate lookup tables

- Now, we're going to populate the lookup tables, using the output of the techniques found in  [PR 1282](https://github.com/cityofaustin/atd-vz-data/pull/1282). Please refer to this PR for context around these commands.

```bash
# drop these schemata, although they won't be here if you're running down this list normally. 

docker exec -i \
-e PGHOST=localhost \
-e PGDATABASE=atd_vz_data \
-e PGPASSWORD=visionzero \
-e PGUSER=visionzero \
visionzero-postgis \
psql -c "drop schema if exists cris_lookup cascade"

docker exec -i \
-e PGHOST=localhost \
-e PGDATABASE=atd_vz_data \
-e PGPASSWORD=visionzero \
-e PGUSER=visionzero \
visionzero-postgis \
psql -c "drop schema if exists vz_lookup cascade"
```

## Get lookup data derived from CRIS spreadsheet that defines the schemata

```bash
# sequences
curl https://gist.githubusercontent.com/frankhereford/906cf1a4d4b6bab90c336d674982b3aa/raw/523ad9bb136588387f7e8e4450900f6c91ba2cfa/global_lookup_sequences.sql | \
docker exec -i \
-e PGHOST=localhost \
-e PGDATABASE=atd_vz_data \
-e PGPASSWORD=visionzero \
-e PGUSER=visionzero \
visionzero-postgis \
psql -v ON_ERROR_STOP=1

# tables
curl https://gist.githubusercontent.com/frankhereford/906cf1a4d4b6bab90c336d674982b3aa/raw/523ad9bb136588387f7e8e4450900f6c91ba2cfa/lookup_table_data.sql | \
docker exec -i \
-e PGHOST=localhost \
-e PGDATABASE=atd_vz_data \
-e PGPASSWORD=visionzero \
-e PGUSER=visionzero \
visionzero-postgis \
psql -v ON_ERROR_STOP=1

# materialized views
curl https://gist.githubusercontent.com/frankhereford/906cf1a4d4b6bab90c336d674982b3aa/raw/523ad9bb136588387f7e8e4450900f6c91ba2cfa/materialized_views.sql | \
docker exec -i \
-e PGHOST=localhost \
-e PGDATABASE=atd_vz_data \
-e PGPASSWORD=visionzero \
-e PGUSER=visionzero \
visionzero-postgis \
psql -v ON_ERROR_STOP=1
```

## CRIS backfill

- Next comes the really, really long part of the process. We'll be shortening it for this testing, but this can reasonably be expected to take up to 40 hours. Any ideas on how to hasten this process are massively welcome. It'd be a huge win for not only the LDM launch process, but our import script as well.
- We're going to backfill /all/ of CRIS' data into the CRIS fact tables. We're going to need the LDM lookup aware CRIS import, which comes in from that same PR, #1282. This code has been merged into this branch.
  - You can expect this 3 month demonstration back-fill to take 1-2 hours to run, for this step. The size of the file you choose here will define how long the rest of this process will take. This particular export contains crashes from all CRIS eras / schemata, which is helpful for a complete testing of the system's ability to handle each CRIS scheme.

```bash
# become root on the sftp endpoint and copy the example extract
# so that the import script can get at it.

# This example backfill extract contains 3 months of 2023 crashes,
# but it includes crashes in at least 3 CRIS schemata / eras.
cp ~/cris-demo-backfill.zip /home/txdot;
chown txdot:txdot /home/txdot/cris-demo-backfill.zip;
```

- If you want to proceed step by step ..

```bash
 # starting from the root of the repo
cd atd-etl/cris_import

# get a shell on a cris import container
docker compose run -it cris-import bash;

# start up the sshd service, because we're our own bastion locally
/sbin/sshd

# run the cris import script to execute the backfill. 
./cris_import.py

```

- OR, you can let the container's orchestration handle it

```bash
docker compose run -it cris-import
```

## Compute VZ dataset

- Next, we’ll pull over the fact tables from the production VZ database, so that we can compute it against what we have now from CRIS and extract fields which have been changed. These changed data will become the initial contents of the VZ fact table schema’s contents.
- We’re going to need some supporting tables to come over. Let’s handle that first.

```bash
docker exec -i \
-e PGHOST=localhost \
-e PGDATABASE=copy_of_production \
-e PGPASSWORD=visionzero \
-e PGUSER=visionzero \
visionzero-postgis \
pg_dump \
-t public.atd_jurisdictions \
-t public.council_districts \
--no-owner --clean --if-exists --no-acl --disable-triggers | \
perl -pe 's/^(SELECT pg_catalog\.set_config)/-- $1/g' | \
docker exec -i \
-e PGHOST=localhost \
-e PGDATABASE=atd_vz_data \
-e PGPASSWORD=visionzero \
-e PGUSER=visionzero \
visionzero-postgis \
psql --set ON_ERROR_STOP=on
```

- Let’s start with creating a schema to hold the old production fact tables and populate them into it.

```bash
docker exec -i \
-e PGHOST=localhost \
-e PGDATABASE=atd_vz_data \
-e PGPASSWORD=visionzero \
-e PGUSER=visionzero \
visionzero-postgis \
psql --set ON_ERROR_STOP=on \
-c "drop schema if exists production_fact_tables cascade;"

docker exec -i \
-e PGHOST=localhost \
-e PGDATABASE=atd_vz_data \
-e PGPASSWORD=visionzero \
-e PGUSER=visionzero \
visionzero-postgis \
psql --set ON_ERROR_STOP=on \
-c "create schema production_fact_tables;"

docker exec -i \
-e PGHOST=localhost \
-e PGDATABASE=copy_of_production \
-e PGPASSWORD=visionzero \
-e PGUSER=visionzero \
visionzero-postgis \
pg_dump \
-t public.atd_txdot_crashes -t public.atd_txdot_units \
-t public.atd_txdot_person -t public.atd_txdot_primaryperson \
--no-owner --no-acl | \
perl -pe 's/public.atd_txdot_crashes/production_fact_tables.atd_txdot_crashes/g' | \
perl -pe 's/public.atd_txdot_units/production_fact_tables.atd_txdot_units/g' | \
perl -pe 's/public.atd_txdot_person/production_fact_tables.atd_txdot_person/g' | \
perl -pe 's/public.atd_txdot_primaryperson/production_fact_tables.atd_txdot_primaryperson/g' | \
perl -pe 's/^(CREATE TRIGGER.*)/-- $1/g' | \
perl -pe 's/^(ALTER TABLE.*person_audit_log.*)/-- $1/g' | \
perl -pe 's/^(SELECT pg_catalog\.set_config)/-- $1/g' | \
docker exec -i \
-e PGHOST=localhost \
-e PGDATABASE=atd_vz_data \
-e PGPASSWORD=visionzero \
-e PGUSER=visionzero \
visionzero-postgis \
psql --set ON_ERROR_STOP=on
```

- This new data needs to be transformed such that it's using the new lookup table IDs out of the LDM. This takes 30+ minutes to run.
```bash
# get back in your shell in the container that 
# provides the execution environment for this process

docker compose run -it cris-import bash

# start up the sshd service, because we're our own bastion locally
/sbin/sshd

# run the script which finds the differences and records them
python ./transform_existing_data.py
```

- Now that we’ve got the old data and the CRIS data, let’s compute what part of the old data is the VZ data.
  - This will take 1-2 hours. 

```bash
# get back in your shell in the container that 
# provides the execution environment for this process

docker compose run -it cris-import bash

# start up the sshd service, because we're our own bastion locally
/sbin/sshd

# run the script which finds the differences and records them
python ./build_vz_contents.py
```

## Triggers to maintain VZ table rows in sync with CRIS schema's tables 

- Now, with the two sets of tables, each pairing up, one from the VZ fact table schema and one from the CRIS fact table schema, let’s put in some triggers on the VZ tables to make sure we always have a VZ record for each entity. We’ll be running [this file](https://raw.githubusercontent.com/cityofaustin/atd-vz-data/ldm-roundup/atd-toolbox/materials_from_ldm_conversion/triggers_to_maintain_VZ_records_for_fact_tables.sql).


```bash
curl https://raw.githubusercontent.com/cityofaustin/atd-vz-data/ldm-roundup/atd-toolbox/materials_from_ldm_conversion/triggers_to_maintain_VZ_records_for_fact_tables.sql  | \
docker exec -i \
-e PGHOST=localhost \
-e PGDATABASE=atd_vz_data \
-e PGPASSWORD=visionzero \
-e PGUSER=visionzero \
visionzero-postgis \
psql  -v ON_ERROR_STOP=1
```

## Exposing the data through views
* We've built and populated the fact tables in the VZ and CRIS schemata, so it's time to expose these through views which coalesce these together. 
```
# Either use the bash terminal in the container from the previous step, or recreated it

docker compose run -it cris-import bash

# start up the sshd service, because we're our own bastion locally
/sbin/sshd

# run the script which finds the differences and records them
python ./create_fact_table_public_views.py
```

## Drop intermediate products

```
docker exec -i \
-e PGHOST=localhost \
-e PGDATABASE=atd_vz_data \
-e PGPASSWORD=visionzero \
-e PGUSER=visionzero \
visionzero-postgis \
psql --set ON_ERROR_STOP=on \
-c "drop database copy_of_production with (force);"

docker exec -i \
-e PGHOST=localhost \
-e PGDATABASE=atd_vz_data \
-e PGPASSWORD=visionzero \
-e PGUSER=visionzero \
visionzero-postgis \
psql --set ON_ERROR_STOP=on \
-c "drop schema if exists production_fact_tables cascade;"
```


#### Ship list
- [ ] Code reviewed
- [ ] Product manager approved

frankhereford commented on vision-zero:
```diff
@@ -177,7 +177,7 @@ def replicateDb(args):
     drop_database_command = [
         "psql",
-        "--command=DROP DATABASE " + os.environ["POSTGRES_DB"] + "",
+        "--command=DROP DATABASE " + os.environ["POSTGRES_DB"] + " with (force)",
```
I needed to make a commit, any commit, to let me open up this WIP PR, so here's a tiny QOL change to the `vision-zero` program that I'd like to tag-along with this PR.

mddilley commented:
@frankhereford I'm going to start looking at this one today. I saw you mentioned merging https://github.com/cityofaustin/atd-vz-data/pull/1282 would make this easier. Is it possible to point this PR to `ldm-cris-import-with-lookup` as a base?

frankhereford commented:
@mddilley wrote:
> @frankhereford I'm going to start looking at this one today. I saw you mentioned merging #1282 would make this easier. Is it possible to point this PR to `ldm-cris-import-with-lookup` as a base?

Awesome idea Mike - thank you for suggesting that. It's rebased!

mddilley commented:
@frankhereford I've followed the instructions up to the `transform_existing_data.py` script which went great 🎉, but I ran into the following error in the transform script.
```
update production_fact_tables.atd_txdot_crashes set
                        hwy_dsgn_hrt_id = (
                            select id
                            from public.hwy_dsgn_hrt
                            where true
                                and source = 'cris'
                                and cris_id = production_fact_tables.atd_txdot_crashes.hwy_dsgn_hrt_id::integer
                            )
 where production_fact_tables.atd_txdot_crashes.hwy_dsgn_hrt_id is not null
update production_fact_tables.atd_txdot_crashes set
                        base_type_id = (
                            select id
                            from public.base_type
                            where true
                                and source = 'cris'
                                and cris_id = production_fact_tables.atd_txdot_crashes.base_type_id::integer
                            )
 where production_fact_tables.atd_txdot_crashes.base_type_id is not null
Traceback (most recent call last):
  File "/app/./transform_existing_data.py", line 224, in <module>
    main()
  File "/app/./transform_existing_data.py", line 143, in main
    cursor.execute(sql)
  File "/usr/local/lib/python3.10/dist-packages/psycopg2/extras.py", line 236, in execute
    return super().execute(query, vars)
psycopg2.errors.InvalidTextRepresentation: invalid input syntax for type integer: ""
```
I'm going to debug some, but I wanted to preserve that before I lose it in my terminal.

mddilley commented:
Small fix for the unit lookup table config in `lookup_table.py` that came out of running the transform script - line 347 has `trlr_type_id` but only `trlr1_type_id` and `trlr2_type_id` columns exist in the unit table. I think that one flew under the radar in the lookup PR.

frankhereford commented:
@mddilley & @roseeichelmann 

I wanted to let you know that I am working on addressing the comments that you both shared with me regarding finding additional fields that suffer from the FK values being stored in character fields with some non-conforming values for a handful of records in our fact tables. 

I have generalized the solution for handling those in the transform script, and I am continuing to try to run it to get it just right. I'll follow up with y'all here, or in the LDM standup meeting, with my progress. 

More than anything, I wanted to let y'all know I'm working on it, and I'll have something ASAP.

frankhereford commented:
@roseeichelmann & @mddilley - thanks for the extra time to sort out those problematic columns in the transform script. I think these surfaced here and not in the previous PR because the previous PR used a pre-generated script to do the transformations, and here they are being done dynamically based on that spreadsheet from CRIS. Here's a comment in slack about picking back up on testing: https://austininnovation.slack.com/archives/C05F2052S67/p1695153386974929?thread_ts=1695149554.951369&cid=C05F2052S67. Thanks!!

mddilley commented:
@frankhereford Happy to report that I made it through the transform script! 🎉 I ran out of resources a couple times so that held me up a bit, but I was good after bumping my allocation in Docker Desktop. I'll keep this process going and pick back up on code review on Monday.

mddilley commented on atd-etl/cris_import/create_fact_table_public_views.py:
```diff
@@ -0,0 +1,584 @@
+    db.execute("drop view if exists public.atd_txdot_crashes;")
+    pg.commit()
+
+    # ldm.atd_txdot_crashes columns we want to compute on the fly
```
Could you expand on this comment some? Might be an artifact from previous iterations.

frankhereford replied:
I've removed that comment. If anything, it's more detritus left over from trying to figure out what to do with what fields, and I think [my previous comment](https://github.com/cityofaustin/atd-vz-data/pull/1288#discussion_r1337797768) outlines my preferred way to handle getting all the fields correct.

mddilley commented on atd-etl/cris_import/create_fact_table_public_views.py:
```diff
@@ -0,0 +1,584 @@
+                    coalesce(vz_facts.atd_txdot_crashes.last_update, cris_facts.atd_txdot_crashes.last_update)
+            end as last_update
+            """ 
+        elif column["vz_column_name"] == "last_update":
```
Should I expect to see `last_update` in the public crash view? I see that field in both the `cris_facts` and `vz_facts` tables but not in the view. It is in `public.atd_txdot_primaryperson` but not in the person, crash, or unit views.

frankhereford replied:
🙈 Oh, thank you Mike! I had totally forgotten to push the `definition` for that column onto the array that later is munged into the VIEW's source. `columns.append(definition)` has been added for crashes. 

But also, while looking at this, it looks like I have duplicated branches to handle `last_update` in this script for 3 of the 4 tables and it's entirely missing from the 4th. I've pushed up a change that addresses these duplications. I /added/ the custom `last_update` handling to the primary person table's special handling. 

Here's [the commit](https://github.com/cityofaustin/atd-vz-data/pull/1288/commits/3bcd68d103737e62182db3f0d196900df60b16d1) that does this, but it doesn't scream off the page exactly what I did. I had two 'elif' blocks for `last_update` and the first one didn't have the `append()` call, so I combined them, and now each main fact table is getting the `last_update` defined correctly and recorded in the view by appending to the `columns` array.

mddilley replied:
Cool, thanks for fixing!

mddilley commented on atd-vzv/src/views/summary/CrashesByMode.js:
I think that `master` may need to be pulled all the way up through the branches that lead to this PR so that these types of changes will be removed from this diff. It seems like it is a little ahead of the branch its base branch.

frankhereford replied:
I think you're right. This PR is built off `ldm-cris-import-with-lookup` which has been hanging out in review for a month now. I'm going to merge master into it, and from there into here, and I'll also merge it into `ldm-collector` which is probably a little dusty. This, and the myriad of other unrelated files that show up in this diff, I believe, are in the same boat. I'll do that integration now and see where it stands. If it doesn't clear it up, I'll pick this thread back up.

frankhereford replied:
This worked out - thanks again for this suggestion. This PR and its antecedent should be much easier to read now. Thanks!

mddilley reviewed (commented):
Added some comments, and I've explored the LDM local DB on my machine quite a bit. I can look at the cris and vz tables alongside the view that unions them, and it is behaving just like advertised. 🚀  Having this built up is going to be so helpful for the trigger work. 🙏

frankhereford commented:
So many good questions @mddilley! I am going to work through them - thank you for these!

frankhereford commented on atd-etl/cris_import/build_vz_contents.py:
```diff
@@ -0,0 +1,559 @@
+
+    SFTP_ENDPOINT_SSH_PRIVATE_KEY = secrets["sftp_endpoint_private_key"]
+
+    compute_for_crashes() # this works - save the 80 missing crashes..
+    compute_for_units() # this works - save for the 164 missing units and the one special case
+    compute_for_person() # this works - save for the 85 missing persons
+    compute_for_primaryperson() # this works - save for the 151 missing primary persons
```
I'm going to remove the comments. If anything, they were notes-to-self from development a good while back. These numbers refer to entities that were found in the CRIS full-reimport, but were missing from our VZDB. IIRC, almost all of the entities stemmed from a small, 1-2 period were we had totally missed data from CRIS for some unknown reason.

However, in the way things are working in this PR, the `build_vz_contents.py` script knows to just create the required record with all `null` field values if it finds it missing. This happens thanks to the new `if public is None:` blocks found in the `compute_for_****` functions.

frankhereford commented on atd-etl/cris_import/build_vz_contents.py:
```diff
@@ -0,0 +1,559 @@
+                keys = ["crash_id"]
+                values = [cris["crash_id"]]
+                for k, v in cris.items():
+                    if (k in ('crash_id', 'council_district', 'in_austin_full_purpose')): # use to define fields to ignore
```
We spoke about this today, and I don't know if we came to a full resolution on what we should do. I think we should go the way that makes writing the triggers easier, and if all else is equal, I think we evaluate them on a case-by-case basis. 

My gut is to get this PR in, even if it's handling almost all fields "generically" and we can create an issue that is limited in scope to audit and choose exactly how we're going to handle each field non-trivial field. This PR can be a scaffold and starting place for us to do that work. Essentially, I want to avoid packing as much LDM work as we can into this already huge PR, and start to break out refinements to the system systematically and intentionally. 

@patrickm02L - I want to tag you since I'm proposing a planning-like operation, and I want to make sure you're 👍 or give you a chance to offer direction if you want.

frankhereford replied:
Cool, I'm going to get one written up today and put it in the epic. Thanks!

frankhereford replied:
Here's an issue describing this work to refine the special handling of columns in the unifying view.

https://github.com/cityofaustin/atd-data-tech/issues/14008

frankhereford commented on atd-etl/cris_import/create_fact_table_public_views.py:
```diff
@@ -0,0 +1,584 @@
+    for column in db:
+        print("Column: ", column["vz_column_name"])
+        if column["vz_column_name"] == "longitude_primary":
+            columns.append("ST_X(COALESCE(vz_facts.atd_txdot_crashes.position, cris_facts.atd_txdot_crashes.position)) as longitude_primary")
```
Everything you've mentioned here is right on the money. I hope we use the dynamic calculation of derived information wherever possible in the views, such as the `ST_X()`s and `ST_Y()`s, since they are so cheap, and then fully leverage the use of DB triggers to maintain fields that are too expensive to compute on an _ad hoc_ basis. It'll be a column by column call on each, perhaps with use of `EXPLAIN ANALYZE` to put some numbers to the cost of the computations. Again, I think this would be work in that [aforementioned issue here](https://github.com/cityofaustin/atd-vz-data/pull/1288#discussion_r1337797768).

mddilley replied:
Cool, yes, after talking about this in the standup meeting, I totally agree that keep this view slim is the way to go. 💯

frankhereford commented on atd-etl/cris_import/transform_existing_data.py:
I'll do my best, but I'd love to have a conversation in a slack huddle to make sure I'm doing your question justice. 

There are two sets of fact tables which need to have all of their foreign key columns updated to reflect the IDs we're using in our combined lookup table system. The first set is the existing set of VZ data fact tables, and this is a one-time operation being handled by this script. The second set of fact tables that need to be updated is the fact tables we get in every CRIS export.  We get a new set of these every time we do an import, and these need to be transformed as the first step of that import process. The code to do this is in the CRIS import ETL process via the branch that this one is based off of. 

The reason that they need to be transformed, beyond not breaking the joins between fact tables and lookup tables, is that we are comparing what we're getting from CRIS and what we have on file. Without that transformation, all or nearly all of the FK values would appear to be changed during a CRIS import or during the `build_vz_contents.py` process. If both fact tables being compared are not using the same lookup tables and those lookup tables' ID values, we don't have a way to do a meaningful comparison, at least not easily. 

All in all, this script is handling the one-time-conversion to the new lookup table IDs for our existing data. 

Please let me know what other questions you may have about this, and I'd be eager to join in a conversation about it if that would be helpful.

mddilley replied:
That is super helpful! Thank you. I'm starting to think that there are some topics that are worth creating some documentation and/or diagrams to spell out. This feels like one of them. I find it easy to get what happens during the LDM transformation (which won't happen daily after the LDM release) and what happens during the import which will happen daily after release.

mddilley reviewed (approved):
Thanks again for putting this and the lookup PR together. It really helped get topics of discussion and future bits of work out in the open. 🙏 I've been able to test the core functionality of the LDM and see it work, and this will give a good base for future PRs. 🚀

roseeichelmann commented:
Yayy I finished running all these commands and have been exploring the new LDM DB that resulted from this testing! :sparkles: 

Below are some things I noted which shouldn't be addressed in this PR, just wanted to point them out before I forget!

I've been comparing the `vz_facts` and `cris_facts` `atd_txdot_crashes` tables and a few things stuck out to me. I'm going to use `crash_id` 13707707 as an example. In the `cris_facts` table, the `in_austin_full_purpose` column has a value of `false`, though that's not actually a value we get from cris. I would imagine that column would maybe only even exist in the `vz_facts` table, or if we decide to have the exact same shape for the `vz_facts` and the `cris_facts` tables, then it should be null in `cris_facts` and that `false` should be in the `vz_facts`. 

Also, for example, with the column `temp_record` for `crash_id` 13707707, both tables have `false` in that column. Once again this is a column we created, so should values for that column only exist in the `vz_facts` table?

Other example columns where the same thing happens are apd_confirmed_fatality, apd_human_update, est_comp_cost, and many more.

I know we decided to address these questions in this new issue you created https://github.com/cityofaustin/atd-data-tech/issues/14008 so I'm just commenting about this here to show other examples where we need to think about how these columns should behave/where should they be.

roseeichelmann reviewed (approved):
This was a monstrous amount of work, thanks for putting it together Frank! Like Mike said this will be really helpful for the trigger PRs and for helping all of us wrap our brains around the LDM :rocket: :face_with_spiral_eyes: 

 I think my only suggestion is to maybe write some quick header comments at the top of some of the python programs briefly describing what they do? Like for the cris_import script, build_vz_contents, parse_cris_schema, etc. Personally I think it's nice when I can quickly read a summary of what the program does.

patrickm02L reviewed (approved):
Amazing work! 💯
